!!! FEATURE: Content Repository Privileges

bwaidelich opened the pull request:
Introduces basic security to the event sourced content repository

This is a breaking change mainly because it requires a new, explicit, workspace role assignment in order for the `live` workspace to be publicly visible:

```shell
./flow workspace:assignrole live "Neos.Flow:Everybody" viewer
```

Otherwise youll face this error:

> Read access denied for workspace "live": User is no Neos Administrator and has no explicit role for workspace "live"


Adjustments to `VisibilityConstraints`:

The method `VisibilityConstraints::frontend()` was renamed to `VisibilityConstraints::default()`.
But for the cases you have used `VisibilityConstraints` at all - as mandatory argument for `getSubgraph` you can now simplify that to use `getContentSubgraph` directly:

_before_

```php
$subgraph = $contentRepository->getContentGraph($nodeAddress->workspaceName)->getSubgraph(
    $nodeAddress->dimensionSpacePoint,
    $nodeAddress->workspaceName->isLive() ? VisibilityConstraints::frontend() : VisibilityConstraints::withoutRestrictions()
);
```

_after_

```php
$subgraph = $contentRepository->getContentSubgraph($nodeAddress->workspaceName, $nodeAddress->dimensionSpacePoint);
```

Note that `getContentGraph(...)->getSubgraph(...)` still works but is only meant to be used if your you want to determine the `VisibilityConstraints` for some case differently.

## Usage

### Workspace Permissions

The following [workspace roles](https://github.com/neos/neos-development-collection/blob/f48b564f4d4cfb18b16dc2485347aa67f4de5836/Neos.Neos/Classes/Domain/Model/WorkspaceRole.php) exist:
- `VIEWER`: Can read from the workspace
  - By default, everybody is a `VIEWER` to the `live` workspace
- `COLLABORATOR`: Can read from and write to the workspace
- `MANAGER`: Can read from and write to the workspace and manage it (i.e. change metadata & role assignments)

Furthermore, personal workspaces can be _owned_ by a particular user. The workspace owner implicitly has the `MANAGER` role, i.e. has full access to that workspace.

> [!NOTE]  
> Workspace roles are transitive. That means that `COLLABORATOR` has all privileges of the `VIEWER` and `MANAGER` has all privileges of the `COLLABORATOR`

To publish changes to a workspace, the authenticated user has to have at least the `VIEWER` role on the workspace to publish and `COLLABORATOR` role on the target workspace.

Workspace role assignments can be changed through the workspace module (WIP, see #5132) or via CLI:

```shell
./flow workspace:assignrole
```

### ReadNodePrivilege

In general, the `ReadNodePrivilege` works like it did before Neos 9: It allows to completely hide sub trees of the content repository from a given user group.
But the implementation and configuration format has changed.

This is how a `ReadNodePrivilege` can be configured via `Policy.yaml`:

```yaml
privilegeTargets:
  'Neos\Neos\Security\Authorization\Privilege\ReadNodePrivilege':
    'Neos.Demo:ReadBlog':
      matcher: 'blog'
```

The `matcher` refers to a `SubtreeTag`, so this example would hide all nodes that are tagged with `blog` (including their descendants) from all users that don't have a `GRANT` on that specific privilege target.

> [!TIP]
> By default, the privilege is active for all configured content repositories. By prefixing the matcher with `<content-repository-id>:` this can be limited to a specific CR (for example `default:blog`)

### EditNodePrivilege

Likewise the purpose of the `EditNodePrivilege` is like before Neos 9: To restrict certain users from changing nodes (includes creation, setting properties or references, disabling/enabling, tagging and moving, ...) in a given subtree.

The new configuration syntax is similar to the one for `ReadNodePrivilege`:

```yaml
privilegeTargets:
  'Neos\Neos\Security\Authorization\Privilege\EditNodePrivilege':
    'Neos.Demo:EditBlogNodes':
      matcher: 'blog'
```

Again, the `matcher` refers to a `SubtreeTag`, so this example would disallow users without a corresponding `GRANT` permission to edit nodes in the subtree that is tagged `blog`.

> [!NOTE]  
> Both privileges expect a subtree to be tagged. That is not yet possible via the Neos UI or CLI but requires interaction with the PHP API (`$contentRepository->handle(TagSubtree::create(....))`) – we'll provide a way to do that, follow #3732 for updates

## Removed privilege types

The following Content Repository privilege types have been removed because their behavior was inconsistent and/or because they led to (performance) issues: `NodeTreePrivilege`, `CreateNodePrivilege`, `ReadNodePropertyPrivilege`, `EditNodePropertyPrivilege`

We will most probably re-add some of the functionality in a future version, but for now you can implement the [AuthProviderInterface](https://github.com/neos/neos-development-collection/blob/f48b564f4d4cfb18b16dc2485347aa67f4de5836/Neos.ContentRepository.Core/Classes/Feature/Security/AuthProviderInterface.php) in order to enforce custom authorization requirements.

## Examples

### Disallow a role to edit nodes of a subtree

With the following `Policy.yaml`:

```yaml
privilegeTargets:
  'Neos\Neos\Security\Authorization\Privilege\EditNodePrivilege':
    'Neos.Demo:EditBlogNodes':
      matcher: 'blog'
roles:
  'Neos.Neos:Administrator':
    privileges:
      -
        privilegeTarget: 'Neos.Demo:EditBlogNodes'
        permission: GRANT
```
only administrators are allowed to edit nodes in the `blog` subtree.

> [!NOTE]  
> The Neos UI does not currently properly reflect readonly nodes – The inspector will be empty and inline editable fields will be editable – but changing their content will lead to an AccessDenied exception

In order to also _hide_  the uneditable subtree, the following can be added:

```yaml
privilegeTargets:
  # ...
  'Neos\Neos\Security\Authorization\Privilege\ReadNodePrivilege':
    'Neos.Demo:ReadBlogNodes':
      matcher: 'blog'
roles:
  'Neos.Neos:Administrator':
    privileges:
      # ...
      -
        privilegeTarget: 'Neos.Demo:ReadBlogNodes'
        permission: GRANT
```

### Allow a role to only edit within a given subtree

To achieve the opposite of the above, granting a user group to _only_ edit nodes within a subtree, the whole tree must be tagged (i.e. the homepage node, all descendants will inherit the tag).

With all nodes being tagged `demosite` and the blog nodes beeing tagged `blog` *in addition*, the following  `Policy.yaml` will do:

```yaml
privilegeTargets:
  'Neos\Neos\Security\Authorization\Privilege\EditNodePrivilege':
    'Neos.Demo:EditAnyNode':
      matcher: 'demosite'
    'Neos.Demo:EditBlogNodes':
      matcher: 'blog'
roles:
  'Neos.Neos:Administrator':
    privileges:
      -
        privilegeTarget: 'Neos.Demo:EditAnyNode'
        permission: GRANT
  'Neos.Neos:Editor':
    privileges:
      -
        privilegeTarget: 'Neos.Demo:EditBlogNodes'
        permission: GRANT
```

With that, the administrator will be able to edit any node, but editors can only edit nodes within the `blog` subtree.

Related: #3732

bwaidelich committed:
FEATURE: Content Repository Privileges

Related: #3732

bwaidelich committed:
WIP: FEATURE: Content Repository Privileges

bwaidelich committed:
Merge branch '9.0' into feature/3732-cr-privileges-v2

bwaidelich committed:
Fix ContentRepositoryAuthProviderFactory

mhsdesign commented on Neos.ContentRepository.Core/Classes/Projection/ContentGraph/VisibilityConstraints.php:
```diff
@@ -29,7 +29,7 @@
      * @param SubtreeTags $tagConstraints A set of {@see SubtreeTag} instances that will be _excluded_ from the results of any content graph query
      */
-    private function __construct(
+    public function __construct(
```
id favour a static `create` method here to be consistent :)

bwaidelich replied:
Done with 14c9f84fd0b899d84ab17dc837d356f7a5494581

mhsdesign commented on Neos.ContentRepository.TestSuite/Classes/Behavior/Features/Bootstrap/CRTestSuiteRuntimeVariables.php:
```diff
@@ -118,7 +118,7 @@ public function visibilityConstraintsAreSetTo(string $restrictionType): void
         $this->currentVisibilityConstraints = match ($restrictionType) {
             'withoutRestrictions' => VisibilityConstraints::withoutRestrictions(),
-            'frontend' => VisibilityConstraints::frontend(),
+            'frontend' => VisibilityConstraints::default(),
```
here or later we should adjust tests to `default`

> VisibilityConstraints are set to "(withoutRestrictions|frontend)

bwaidelich replied:
Done with ac70e83fd8d9889858586526c87e7aa595c4bb2c

mhsdesign commented on Neos.Neos/Classes/Domain/Service/WorkspaceService.php:
```diff
@@ -44,6 +45,8 @@
 /**
  * Central authority to interact with Content Repository Workspaces within Neos
  *
+ * TODO evaluate permissions for workspace changes
```
todo

mhsdesign commented on Neos.Neos/Classes/Security/Authorization/Privilege/SubtreeTagPrivilege.php:
```diff
@@ -0,0 +1,40 @@
+use Neos\Flow\Security\Exception\InvalidPrivilegeTypeException;
+
+/**
+ * TODO docs
```
tofo

mhsdesign commented on Neos.Neos/Classes/Security/ContentRepositoryAuthProvider/ContentRepositoryAuthProvider.php:
```diff
@@ -0,0 +1,126 @@
+        $user = $this->userService->getCurrentUser();
+        if ($user === null) {
+            return UserId::forSystemUser();
+        }
```
hmmmm

bwaidelich replied:
outdated, see https://github.com/neos/neos-development-collection/blob/feature/3732-cr-privileges-v2/Neos.Neos/Classes/Security/ContentRepositoryAuthProvider/ContentRepositoryAuthProvider.php#L57-L60

mhsdesign commented on Neos.Neos/Classes/Security/ContentRepositoryAuthProvider/ContentRepositoryAuthProvider.php:
```diff
@@ -0,0 +1,126 @@
+            }
+        }
+        return new VisibilityConstraints(
+            SubtreeTags::fromArray($restrictedSubtreeTags)
```
deduplicate array?

bwaidelich replied:
`SubtreeTags` deduplicate themselves, https://github.com/neos/neos-development-collection/blob/9.0/Neos.ContentRepository.Core/Classes/Feature/SubtreeTagging/Dto/SubtreeTags.php#L34-L36

mhsdesign replied:
splendid :D

mhsdesign commented on Neos.Neos/Classes/Security/ContentRepositoryAuthProvider/ContentRepositoryAuthProvider.php:
```diff
@@ -0,0 +1,126 @@
+        // PublishWorkspace
+        // PublishIndividualNodesFromWorkspace
+        // RebaseWorkspace
+        if ($command instanceof CreateWorkspace) {
```
always allow root workspace?

bwaidelich replied:
is that a question?
I would suggest to always _disallow_ `CreateRootWorkspace`, i.e. it needs to be called from CLI or with `$securityContext->withoutAuthorizationChecks(...)`

mhsdesign replied:
jup that makes sense after thinking about it ... and as the live workspace will be already created out of the box (on migration, import, or site create) i think we really dont have to permit the general use here. Its merely a low level thing.

mhsdesign commented on Neos.ContentRepository.Core/Classes/ContentRepository.php:
```diff
@@ -242,9 +249,24 @@ public function resetProjectionState(string $projectionClassName): void
+    /**
+     * @throws WorkspaceDoesNotExist if the workspace does not exist
+     */
+    public function getContentSubgraph(WorkspaceName $workspaceName, DimensionSpacePoint $dimensionSpacePoint): ContentSubgraphInterface
```
We probably need to update the docs.io and also use that in the Neos.Ui ... and in Neos.Neos as well ... maybe we can also introduce just a followup to adjust all usages to use the new method?

bwaidelich replied:
https://github.com/neos/neos-ui/pull/3871 adjusts the UI to use the new API where applicable.

mhsdesign commented on Neos.Neos/Classes/Security/ContentRepositoryAuthProvider/ContentRepositoryAuthProvider.php:
```diff
@@ -0,0 +1,126 @@
+        return UserId::fromString($user->getId()->value);
+    }
+
+    public function getVisibilityConstraints(WorkspaceName $workspaceName): VisibilityConstraints
```
question: should `Neos.Neos:Backend.GeneralAccess` be anywhere evaluated like before?

bwaidelich replied:
Good catch. But I don't think so. That privilege is just used to determine if the user has access to the Neos backend at all. It's not really related to the Content Repository I think

mhsdesign commented on Neos.ContentRepositoryRegistry/Classes/ContentRepositoryRegistry.php:
1. allow `NodeAddress` in subgraph for node? and or also introduce `resolveNodeByNodeAddress`. -> can be a followup
2. is `Node->subgraphIdentity` now obsolet?

mhsdesign replied:
see answer in https://github.com/neos/neos-development-collection/pull/5298#issuecomment-2446886224

bwaidelich committed:
Simplify `AuthProviderInterface::getWorkspacePrivilege()`

bwaidelich committed:
Always grant read access to `live` workspace

bwaidelich committed:
Support CR specific subtree tag privileges

```yaml

privilegeTargets:

  'Neos\Neos\Security\Authorization\Privilege\SubtreeTagPrivilege':

    'Some.Package:FooInAllContentRepositories':
      matcher: 'foo'

    'Some.Package:BarInDefaultContentRepository':
      matcher: 'default:bar'
```

bwaidelich committed:
Merge branch '9.0' into feature/3732-cr-privileges-v2

# Conflicts:
#	Neos.Neos/Classes/View/FusionExceptionView.php

bwaidelich committed:
Rename `AuthProviderInterface::getUserId()`

to `getAuthenticatedUserId()` and make return type nullable

bwaidelich committed:
Extract permission evaluation from `ContentRepositoryAuthProvider` to new `ContentRepositoryAuthorizationService` singleton

mhsdesign commented on Neos.Neos/Classes/Domain/Service/WorkspaceService.php:
```diff
@@ -262,37 +262,73 @@ public function getWorkspaceRoleAssignments(ContentRepositoryId $contentReposito
+                CASE
+                    WHEN role='MANAGER' THEN 1
+                    WHEN role='COLLABORATOR' THEN 2
+                    WHEN role='VIEWER' THEN 3
```
hmm can we make this duplicated logic of `WorkspaceRole` transparent here by maybe using string manipulation? (im referring to the specificity method)

mhsdesign replied:
thanks for https://github.com/neos/neos-development-collection/pull/5298/commits/1a4629701eab78ec9a7c58da5863e65479f16ae4 ❤️

bwaidelich committed:
Merge branch '9.0' into feature/3732-cr-privileges-v2

bwaidelich committed:
Move auth related classes to `Feature/Security`

and add dedicated `AccessDenied` exception

bwaidelich committed:
Merge branch '9.0' into feature/3732-cr-privileges-v2

bwaidelich committed:
`VisibilityConstraints` not only for Neos users

bwaidelich committed:
Merge branch '9.0' into feature/3732-cr-privileges-v2

bwaidelich committed:
Tweaks

bwaidelich committed:
Fix doc comments for `workspace:assignRole` command

bwaidelich committed:
Determine privileges based on the (authenticated) `Account`

..instead of the Neos `User`.
Reasoning:
- A Neos `User` can have multiple accounts assigned
- It makes sense to evaluate all authenticated roles e.g. for frontend authentication)

bwaidelich committed:
Adjust behat tests to account based ACL

bwaidelich committed:
Merge branch '9.0' into feature/3732-cr-privileges-v2

bwaidelich committed:
Tweak `AuthProvideInterface` method names

and add doc comments

bwaidelich committed:
Ignore disabled authorization checks when evaluating visibility constraints

bwaidelich committed:
Merge branch '9.0' into feature/3732-cr-privileges-v2

bwaidelich committed:
Merge branch '9.0' into feature/3732-cr-privileges-v2

# Conflicts:
#	Neos.ContentRepository.Core/Classes/ContentRepository.php

bwaidelich committed:
FEATURE: EditNodePrivilege!

mhsdesign commented on Neos.Neos/Classes/Command/CrCommandController.php:
```diff
@@ -125,6 +125,7 @@ public function importCommand(string $path, string $contentRepository = 'default
         // set the live-workspace title to (implicitly) create the metadata record for this workspace
         $this->workspaceService->setWorkspaceTitle($contentRepositoryId, WorkspaceName::forLive(), WorkspaceTitle::fromString('Live workspace'));
         $this->workspaceService->assignWorkspaceRole($contentRepositoryId, WorkspaceName::forLive(), WorkspaceRoleAssignment::createForGroup('Neos.Neos:LivePublisher', WorkspaceRole::COLLABORATOR));
+        $this->workspaceService->assignWorkspaceRole($contentRepositoryId, WorkspaceName::forLive(), WorkspaceRoleAssignment::createForGroup('Neos.Neos:Everybody', WorkspaceRole::VIEWER));
```
nice :D

mhsdesign commented on Neos.Neos/Classes/Domain/Model/NodePermissions.php:
```diff
@@ -0,0 +1,60 @@
+ * @api
+ */
+#[Flow\Proxy(false)]
+final readonly class NodePermissions
```
as far as i can see this is just used internally right? do we need this api?

bwaidelich replied:
Good call, changed with abca5bbce0c375c1e4a3400dd97b087a568e7233

mhsdesign commented on Neos.Neos/Classes/Domain/Model/WorkspacePermissions.php:
```diff
@@ -22,29 +23,48 @@
-        public bool $write,
-        public bool $manage,
-    ) {
+    public static function create(
```
hmm maybe a rather `withReason` instead of adding it to constructors but thats a nitpick ... nevermind :D

bwaidelich replied:
a reason is required, that's why I added it to all constructors.
But I made this class `@internal` now and the two remaining usages of this method use named parameters

mhsdesign commented on Neos.Neos/Classes/Domain/Model/WorkspaceRoleAssignment.php:
```diff
@@ -15,34 +15,30 @@
         public WorkspaceRoleSubject $subject,
         public WorkspaceRole $role,
     ) {
     }
```
nice deduplication :)

mhsdesign commented on Neos.Neos/Classes/Domain/Model/WorkspaceRoleAssignments.php:
```diff
@@ -39,7 +38,7 @@ public function isEmpty(): bool
     }
 
-    public function getIterator(): Traversable
+    public function getIterator(): \Traversable
```
❤️

mhsdesign commented on Neos.Neos/Classes/Domain/Model/WorkspaceRoleSubject.php:
```diff
@@ -12,28 +12,47 @@
+        return $this->type === $other->type && $this->value === $other->value;
+    }
+
+    public function __toString(): string
```
isnt this rather a custom toDebugString() method?

bwaidelich replied:
right.. But I can imagine that this is being used to render permission details in some management view (e.g. the workspace / user module in Neos) and it makes things easier to convert this. Accidental type conversion is not realistic in this case IMO

mhsdesign replied:
hmm for the workspace module i think were currently going the way of creating dedicated view models that extract the needed information.

see https://github.com/neos/neos-development-collection/blob/eedfc59964e3bbce21a68a27a2df1f02aac3f58f/Neos.Workspace.Ui/Classes/ViewModel/RoleAssignmentListItem.php

bwaidelich replied:
I just removed it with 3c6ab69eee4599b1e985ee8c966adbcd237485da :)

mhsdesign commented on Neos.Neos/Classes/Domain/Service/WorkspaceService.php:
```diff
@@ -80,6 +79,8 @@ public function getWorkspaceMetadata(ContentRepositoryId $contentRepositoryId, W
     /**
      * Update/set title metadata for the specified workspace
+     *
+     * NOTE: The workspace privileges are not evaluated for this interaction, this should be done in the calling side if needed
```
hmmm but why not :D I guess to avoid zick zack?

bwaidelich replied:
right.. I mean we could inject the AuthService here and evaluate, but it creates a weird dependency graph and since this is a rather low level method (and no high risk) I went for the easy way out :)

mhsdesign replied:
hmm yes i agree ...

but regarding 

> since this is a rather low level method

just so we are on the same side the whole `WorkspaceService` is _THE_ way of working with workspace ... the CR workspace commands are low level :D ... not everything can be low level :D there needs to be public parts xD

mhsdesign replied:
and that we cannot wire it together does show that there is a little flaw somewhere hidden ... but thats fine :D

bwaidelich replied:
low level as in: not required for the _regular_ interaction with a CR.
I agree that it at least feels wrong not to have the ACL enforced when changing the title/description metadata of a workspace but I don't think that it is an issue in reality and that the bidirectional dependency just to solve this case would be worse. But maybe others have a different opinion!?

mhsdesign replied:
cc @kitsunet

kitsunet replied:
I am torn here between "oh my, we put so much on the shoulders of the end user with all this" and "yes, this is modular and flexible to use". I think I lean towards doing it like this and putting it on the implementer to decide if they want/need protection in the interaction. I just wonder because that makes for a weird collection of methods in the API, what about the ones that do _not_ explicitly state the are not protected, like e.g. getPersonalWorkspaceForUser

Maybe we should put this up as a general info/warning that nothing in this class is protected and you should do the respective checks before using it's APIs?

bwaidelich replied:
> I just wonder because that makes for a weird collection of methods in the API

Yes, I agree.
I mean: AFAIS it's really "just" about these two methods, `setWorkspaceTitle` and `setWorkspaceDescription` as they are no longer a CR core concern (tbh: maybe we should have _kept_ it one..).

> what about the ones that do not explicitly state the are not protected

In the current implementation there is no restriction to _reading_ workspaces, everybody can read all information about all workspaces (as long as the UI allows them).

I think we can justify the current state with this argumentation: The CR core ensures the _important_ protections (which user can manage workspaces, read/edit nodes from certain workspaces) and it's the responsibility of the Application/UI layer to hide the rest – but it would not mean a security breach if they didn't..

> Maybe we should put this up as a general info/warning that nothing in this class is protected

Yes, good point. I'll add something like the above statement to the class

mhsdesign replied:
Fixed via

- [TASK: Split mighty WorkspaceService into `WorkspaceMetadataAndRoleR…](https://github.com/neos/neos-development-collection/pull/5298/commits/952932700344fbe63c1a516404582c1a054433ea) 
[9529327](https://github.com/neos/neos-development-collection/pull/5298/commits/952932700344fbe63c1a516404582c1a054433ea)
- [FEATURE: Check permissions for operations inside WorkspaceService](https://github.com/neos/neos-development-collection/pull/5298/commits/4f07965b295b93d105214bac67187ec893b34a24)

the `WorkspaceService` got to bloated and now the `WorkspaceService` and the Cr Auth Service use a low level INTERNAL `WorkspaceMetadataAndRoleRepository` for this task. That allows us to require the Cr Auth Service _inside_ the `WorkspaceService` without any lazy object hacks.

The reason why bastian and me rediscussed this was because well `setWorkspaceTitle` is not thaaat important but having security for `assignWorkspaceRole` would be nice and it now is that way :)

mhsdesign commented on Neos.Neos/Classes/Security/Authorization/ContentRepositoryAuthorizationService.php:
```diff
@@ -0,0 +1,218 @@
+        );
+    }
+
+    private function nodeForNodeAddress(NodeAddress $nodeAddress): ?Node
+    {
+        return $this->contentRepositoryRegistry->get($nodeAddress->contentRepositoryId)
+            ->getContentGraph($nodeAddress->workspaceName)
+            ->getSubgraph($nodeAddress->dimensionSpacePoint, VisibilityConstraints::withoutRestrictions())
+            ->findNodeById($nodeAddress->aggregateId);
+    }
```
this is a bit cheesy that we pass the node address from around (in the expectation that its cheap) and internally we still resolve the node?

bwaidelich replied:
Mh. The reason is the `ContentRepositoryAuthProvider` that needs to determine the node permissions for a given command (see https://github.com/neos/neos-development-collection/blob/0bf97e2779edc863178ac1882a0c6328fa5a62b5/Neos.Neos/Classes/Security/ContentRepositoryAuthProvider/ContentRepositoryAuthProvider.php#L93-L100)

Do you have a better idea, i.e. should the `ContentRepositoryAuthProvider` get the `ContentRepository` instance injected to resolve the node itself?

bwaidelich replied:
@mhsdesign I think this is better now: https://github.com/neos/neos-development-collection/pull/5298/commits/fedfac7da910fd1886aed2327bb9d27aca946966 thanks for the pointer!

mhsdesign replied:
Jup its already better thanks ❤️ now one can see that ist not cheap :D

One more thing though if i may :) i would prefer to pass the `ContentGraphReadModelInterface` here as the pattern of passing the content repository registry to such low level extension points is imo super dirty and creates an odd dependency graph and has the danger of recursion if improperly used.

see https://github.com/neos/neos-development-collection/pull/5339

bwaidelich replied:
Good point, but how would you make the `ContentGraphReadModel` available? Another ugly CR service or do we need some `ContentRepositoryRegistry::getContentGraphReadModel()`??

mhsdesign replied:
via the `ContentRepositoryAuthProviderFactory` inside -> `build` as argument id say. I think we should NOT promote the pattern of fetching the `ContentRepositoryRegistry` for those core things!

mhsdesign replied:
via the `ContentRepositoryAuthProviderFactory` inside -> `build` as argument id say. I think we should NOT promote the pattern of fetching the `ContentRepositoryRegistry` for those core things!

bwaidelich replied:
Makes sense, but the `ContentRepositoryRegistry` doesn't have an instance of the `ContentGraphReadModelInterface`

mhsdesign replied:
Ahh is it build too earlly? I think the `ContentRepositoryAuthProviderFactory` must be passed to the CR. factory and there called ->build(CrId, CRReadModel, options)

bwaidelich replied:
Finally, b4e0c7829b61bc962986a898fcc10d929d0632a6 should be better :)

mhsdesign replied:
hehe thanks :D Perfekt!

mhsdesign commented on Neos.Neos/Classes/Security/Authorization/Privilege/EditNodePrivilege.php:
```diff
@@ -0,0 +1,28 @@
+use Neos\Flow\Security\Exception\InvalidPrivilegeTypeException;
+
+/**
+ * TODO docs
```
todo:)

bwaidelich replied:
Done with 187b80ea94840e5e75cc5162c22f6cd50738ec20

mhsdesign reviewed (commented):
Thank youuuuu for the extensive work and love you put into this ❤️ 

This is my first batch of comments that im finally submitting. Even though i am part of this journey i think i still have to dig deeper and test it :) Sooo this is just a first review.

And a further question: (maybe out of scope but still) does this mean we can deprecate the ugly `VisibilityConstraints` property on the node? Fetching would always default to the current users and id say that would be perfect?
And the `VisibilityConstraints` are not part of the nodes identity (see node address) and might just cause confusion?
If you really need them you can ask the subgraph ... except maybe not? See the frontend rendering case? Global state is never cool :D

mhsdesign committed:
TASK: Dont expose `$visibilityConstraints` in `SiteNodeUtility`

mhsdesign committed:
TASK: Use current user visibility constraints for data source controller

bwaidelich committed:
Merge branch '9.0' into feature/3732-cr-privileges-v2

bwaidelich committed:
Introduce named `VisibilityConstraints` constructor

bwaidelich committed:
Adjust Behat tests to refer to Visibility constraints "default" instead of "frontend"

bwaidelich committed:
Mark `NodePermissions` and `WorkspacePermissions` internal

bwaidelich committed:
Deduplicate workspace role specificity determination

bwaidelich committed:
Merge branch '9.0' into feature/3732-cr-privileges-v2

bwaidelich committed:
Tweak resolution of current node in `ContentRepositoryAuthProvider`

bwaidelich commented:
> does this mean we can deprecate the ugly VisibilityConstraints property on the node? Fetching would always default to the current users and id say that would be perfect?

We need to keep that property unfortunately, because it is used to transport the "context" the node is accessed with.
I.e. if you pass a node without any restrictions into a Fusion View, all code parts below (Eel helpers, FlowQuery operations, ...) have to respect that

bwaidelich committed:
Revert `RebasableToOtherWorkspaceInterface` extension

and check against concrete command classes

bwaidelich committed:
Fix workspace permissions check

mhsdesign commented on Neos.Neos/Classes/Security/Authorization/ContentRepositoryAuthorizationService.php:
```diff
@@ -0,0 +1,200 @@
+    /**
+     * Determines the {@see WorkspacePermissions} the given user has for the specified workspace
+     */
+    public function getWorkspacePermissionsForAccount(ContentRepositoryId $contentRepositoryId, WorkspaceName $workspaceName, Account $account): WorkspacePermissions
```
this was the discussion `User` vs `Account` you started in [slack](https://neos-project.slack.com/archives/C04PYL8H3/p1729670810486759) right?

Copied here for reference and as explains _why_ we did it :D

---------------

There will be a [ContentRepositoryAuthorizationService](https://github.com/neos/neos-development-collection/blob/cab0bd6bca905000b35b9e16d80eefcda1d8c5c6/Neos.Neos/Classes/Security/Authorization/ContentRepositoryAuthorizationService.php) that has public methods like

```php
getWorkspacePermissionsForUser($crId, $workspaceName, User $user): WorkspacePermissions {...}
```

But this only allows to get workspace permissions for an authenticated backend user – it would make sense to allow this for any authenticated account IMO
For visibility constraints I already adjusted it to
```php
getVisibilityConstraintsForRoles(array $roles): VisibilityConstraints {...}
```

but this untyped roles array is also not very nice.
Besides for workspace permissions we also need to check whether the user is the owner of that workspace, so roles would not be enough..
I though about using `Account` in those places but that also has some drawbacks:
- We'll have to expand the roles (similar to how we do it in the [UserService](https://github.com/neos/neos-development-collection/blob/9.0/Neos.Neos/Classes/Domain/Service/UserService.php#L736-L759)) even though `Security\Context::getRoles()` will mostly already have those "expanded" roles
- A Neos user can be assigned to multiple accounts – it would not be easily possible to determine workspace permissions for a user.. Thinking about it, we couldn't really do that before because `UserService::getAllRoles()` will evaluate all linked accounts, but only one will be authenticated at a time
- Lastly: The workspace owner is currently the persistent identifier of the neos `User` – so to determine whether an account is the owner, we'd have to do sth like `UserService::getUser($account->getAccountIdentifier())`

mhsdesign replied:
Do we want to keep the `expandAccountRoles` roles logic here or could that be moved to flow, be cached, and possibly reused by the security context?

Also whenever we duplicate logic like `neosUserFromAccount` i would plead for adding documentation where it comes from or use the actual service here like the `UserService` no?

bwaidelich replied:
> Do we want to keep the expandAccountRoles roles logic here

I would say it's an ok solution _for now_. We could centralize that logic to Flow in a separate PR, but that needs some more thought – especially if it adds runtime caches.
Also with the current API it would need to be sth like `PolicyService::expandRolesForAccount()` or similar (or we expand the `Account` model to already contain the expanded roles) because otherwise we would have to pass that untyped `$roles` array from the outside

> Also whenever we duplicate logic like neosUserFromAccount i would plead for adding documentation

The implementation is a one-liner basically:
```php
$this->partyService->getAssignedPartyOfAccount($account)
```
and that's just a core feature of the Party package. But I'm always in favor of more docs. Where would you suggest to add them and in what way?

bwaidelich replied:
@mhsdesign can we mark this one resolved for now?

mhsdesign replied:
this is more of a FYI for other reviewers so that can stay open for ever :)

mhsdesign commented on Neos.Workspace.Ui/Classes/Controller/WorkspaceController.php:
```diff
@@ -1013,7 +1027,7 @@ protected function prepareBaseWorkspaceOptions(
         WorkspaceName $excludedWorkspace = null,
     ): array {
-        $user = $this->userService->getCurrentUser();
+        $authenticatedAccount = $this->securityContext->getAccount();
```
As we switch from `getCurrentUser` to `getAccount` and then ask the party service, i just want to highlight that this is _NOT_ exactly the same.

In `getAccount` we check if the token `isAuthenticated` and return the first account.

In the user service we would loop though all `getAuthenticationTokens` and get the user for the first account that has one.

I dont exactly understand why there might be multiple accounts but its documented like this:
 
> Note: There might be more currently authenticated account in the remaining tokens

bwaidelich replied:
That's right. The possibility to attach _multiple_ accounts to a single Neos user was added sometimes in the past. But I doubt that it is used _anywhere_ in practice because it has some fundamental logic flaws.
For example: `UserService::getAllRoles()` returns all roles of all attached accounts. But there is always only one authenticated account at a time, so that combination of roles does not really help..

mhsdesign replied:
cc @kitsunet

kitsunet replied:
This is just half the problem, I think the more possible scenario is that you have frontend users and logged in with one of those AND your backend account. I this case you want to ensure in places like this here that you get the right account. Yes I know people will _probably_ not have their frontend users in the request patterns for backend, but I have seen it and I think it would get hard to debug and understand if it goes wrong even if that is a rare case.
tl;dr; I really think we should keep the existing logic.

bwaidelich replied:
What would the existing logic be in this case?

kitsunet replied:
Iterated authenticated accounts and find the first that has a neos user attached

bwaidelich replied:
> Iterated authenticated accounts and find the first that has a neos user attached

I see. I was afraid that you'd suggest to change the whole API.
Unfortunately the security context does not allow to iterate authenticated accounts, but what about 8b8eea7700669e1edaece155d6ebbfc60a39022a ?

mhsdesign replied:
Uiuiuiuuiu what have we gotten ourselves into 😭 😅 ... maybe that direction?

But you forgot to change it at all places: https://github.com/neos/neos-development-collection/blob/41e6b57ba8917ef394e97e4ab511db37453032d6/Neos.Neos/Classes/Security/ContentRepositoryAuthProvider/ContentRepositoryAuthProvider.php#L202

mhsdesign replied:
Thanks basti for simplifying and fixing this via https://github.com/neos/neos-development-collection/pull/5298/commits/466c89e583dbbc9bfdf356e256f7f5cea69ec28f ❤️

mhsdesign commented on Neos.Workspace.Ui/Classes/Controller/WorkspaceController.php:
```diff
@@ -156,9 +161,9 @@ classification: $workspaceMetadata->classification->name,
-            throw new \RuntimeException('No user authenticated', 1720371024);
+        $authenticatedAccount = $this->securityContext->getAccount();
+        if ($authenticatedAccount === null) {
+            throw new AccessDeniedException('No user authenticated', 1720371024);
```
`AccessDeniedException` is probably correct but it would seem like _that_ is flows way of security ... this should never happen no?

bwaidelich replied:
That's right, it's just a type-safety thing because `Security\Context::getAccount()` is nullable, but it should not happen in practice™

mhsdesign replied:
With the new nullability i could actually imagine to just ignore that check here and use `$currentUser?->getId()` like in other places  ... that makes the code imo less quirky as the security part in flow is defined via yaml and not in the code.

But will do that inside of sebs workspace pr instead of messing with the workspace controller to much :D

mhsdesign commented on Neos.Neos/Classes/Security/ContentRepositoryAuthProvider/ContentRepositoryAuthProvider.php:
```diff
@@ -0,0 +1,207 @@
+    private function requireWorkspacePermission(WorkspaceName $workspaceName, string $permission): Privilege
+    {
+        $workspacePermissions = $this->getWorkspacePermissionsForCurrentUser($workspaceName);
+        if (!$workspacePermissions->{$permission}) {
```
that trick has to be documented please 😅 i think when telling phpstand that `$permission` MUST be a constant this is event statically analysable...

bwaidelich replied:
it's just a private method, but I agree. I have split it into two methods with 0571abf6fec480ef08ae9352cf1875c8dfe3f91e

mhsdesign replied:
jup but these magic always comes with the pain of one does not find its usages ... and phpstand is totally blind :D Thanks!

mhsdesign commented on Neos.Neos/Classes/Security/ContentRepositoryAuthProvider/ContentRepositoryAuthProvider.php:
```diff
@@ -0,0 +1,207 @@
+        }
+
+        /** @var NodeAddress|null $nodeThatRequiresEditPrivilege */
+        $nodeThatRequiresEditPrivilege = match ($command::class) {
```
should we maybe extract all this node based security in a separate method?

or have some comment sections like what we are checking where :D

```
#
# Node based command privileges  
# 

...

#
#
```

bwaidelich replied:
I extracted one method with da297ece279536aa9911711d55ffdfce2034c400 – probably there is more potential.. Feel free to amend

mhsdesign replied:
thanks makes it already more readable :D

mhsdesign commented on Neos.Neos/Classes/Security/Authorization/Privilege/SubtreeTagPrivilegeSubject.php:
```diff
@@ -0,0 +1,41 @@
+    ) {
+    }
+
+    public function __toString(): string
```
i think that should be `toDebugString` ... imo `__toString` should only be used for simple vo ...

bwaidelich replied:
I just removed it with 643ce497df50c1f37ab711c35f81442639579c80

mhsdesign commented on Neos.Neos/Classes/Security/Authorization/Privilege/AbstractSubtreeTagBasedPrivilege.php:
```diff
@@ -0,0 +1,78 @@
+    public function getSubtreeTags(): SubtreeTags
+    {
+        $this->initialize();
+        assert($this->subtreeTagsRuntimeCache !== null);
```
```php
    /** @phpstan-assert $subtreeTagsRuntimeCache !== null */
    private function initialize(): void
```

:)

bwaidelich replied:
?

bwaidelich replied:
Ah, I think I got you: da297ece279536aa9911711d55ffdfce2034c400

mhsdesign commented on Neos.Neos/Classes/Security/Authorization/ContentRepositoryAuthorizationService.php:
```diff
@@ -0,0 +1,200 @@
+    /**
+     * Determines the {@see WorkspacePermissions} an anonymous user has for the specified workspace (aka "public access")
+     */
+    public function getWorkspacePermissionsForAnonymousUser(ContentRepositoryId $contentRepositoryId, WorkspaceName $workspaceName): WorkspacePermissions
```
As `Neos.Flow:Anonymous` is documented like

> Likewise, the magic role ``Neos.Flow:Anonymous`` is added to the security context if no user
is authenticated and ``Neos.Flow:AuthenticatedUser`` if there is an authenticated user.

its a little weird to mix the Anonymous and User wording here... maybe it should be named

- `getWorkspacePermissionsForNoUser`

but while writing it i noticed that that doesnt make it any better :D

mhsdesign replied:
Or maybe just `getWorkspacePermissionsForEverybody`?

bwaidelich replied:
"Everybody" would be wrong IMO since it's _not_ everybody, it's the unauthenticated (aka anonymous) user.
To me "AnonymousUser" makes much more sense than "NoUser" in this context. But we can discuss that

mhsdesign commented on Neos.ContentRepository.Core/Classes/Feature/Security/Dto/Privilege.php:
```diff
@@ -0,0 +1,38 @@
+{
+    private function __construct(
+        public bool $granted,
+        public string $reason,
```
shouldn't this also be private and `getReason()` here like below?

bwaidelich replied:
What "below" are you referring to?

mhsdesign replied:
the `WorkspacePermissions`

bwaidelich replied:
Aligned with da297ece279536aa9911711d55ffdfce2034c400

mhsdesign reviewed (commented):
Thanks for the adjustments so far ❤️ i read through all the code again (this time button up) so i think i have seen all of it `*` :D Besides these questions that are left to be discussed this looks super good.

`*` Note that i have not really reviewed fully the new flow privileges and how they are implemented :)

bwaidelich committed:
Merge branch '9.0' into feature/3732-cr-privileges-v2

bwaidelich committed:
Remove `WorkspaceRoleSubject::__toString()`

because it led to [debates](https://github.com/neos/neos-development-collection/pull/5298#discussion_r1818140741)

bwaidelich committed:
Split `ContentRepositoryAuthProvider::requireWorkspaceWritePermission()` into two methods

bwaidelich committed:
Tweak AuthProvider wiring

bwaidelich committed:
Make `Privilege::reason` private

to be in sync with `WorkspacePermissions`

bwaidelich committed:
Remove `SubtreeTagPrivilegeSubject::__toString()`

bwaidelich committed:
Replace PHP assert by phpstan assert in `AbstractSubtreeTagBasedPrivilege`

bwaidelich committed:
Add inline docs

bwaidelich committed:
Extract `ContentRepositoryAuthProvider::nodeThatRequiresEditPrivilegeForCommand()`

bwaidelich commented:
@mhsdesign Thanks so much for your valuable feedback. I took the freedom to mark some of the discussions resolved whenever I thought so. Can you please re-check if there's anything left to do in your opinion?

bwaidelich committed:
TASK: Tests for "Content Repository Privileges"

Related: #3732
Related: #5298

bwaidelich committed:
moar tests

bwaidelich committed:
moar better tests

bwaidelich committed:
Merge branch '9.0' into feature/3732-cr-privileges-tests

kitsunet commented:
Like this a whole lot so far. Some open comments, but I think we are close?!

bwaidelich committed:
Merge branch '9.0' into feature/3732-cr-privileges-v2

# Conflicts:
#	Neos.ContentRepository.Core/Classes/Factory/ContentRepositoryFactory.php

bwaidelich committed:
Extract authenticated account from current Neos user

...instead of the security context

bwaidelich committed:
Fix `test_parallel` CR settings

bwaidelich committed:
Merge branch 'feature/3732-cr-privileges-v2' into feature/3732-cr-privileges-tests

bwaidelich committed:
Simplify `ContentRepositoryAuthorizationService` API

according to review comments

bwaidelich committed:
Fix role conversion in `ContentRepositoryAuthorizationService`

mhsdesign commented on Neos.Neos/Classes/Security/Authorization/ContentRepositoryAuthorizationService.php:
```diff
@@ -34,133 +33,71 @@
+     * @param array<Role> $roles
      */
-    public function getVisibilityConstraintsForAnonymousUser(ContentRepositoryId $contentRepositoryId): VisibilityConstraints
+    public function getNodePermissions(Node $node, array $roles): NodePermissions
```
A sorry my bad `NodePermissions` should be `@api` after all if its returned from this `@api` service? And so `WorkspacePermissions`?

bwaidelich replied:
True, adjusted with 6397ff4ec27caa004cc796122cea44e7bbb8c762

bwaidelich committed:
Mark `NodePermissions` and `WorkspacePermissions` `@api`

bwaidelich committed:
Merge branch 'feature/3732-cr-privileges-v2' into feature/3732-cr-privileges-tests

# Conflicts:
#	Neos.Neos/Tests/Behavior/Features/Bootstrap/WorkspaceServiceTrait.php

mhsdesign committed:
TASK: Split mighty `WorkspaceService` into `WorkspaceMetadataAndRoleRepository`

By moving the db implementation logic in a separate class, the dependency chain is cleaned up, as the `WorkspaceService` would previously hold the cr which would hold the `ContentRepositoryAuthorizationService` which has the `WorkspaceService`

Also we didnt want to add further zickzack and previously it would not have been able to acquire the `ContentRepositoryAuthorizationService` without hacks in the `WorkspaceService` to evaluate permissions for `setWorkspaceTitle` or `assignWorkspaceRole`

mhsdesign committed:
FEATURE: Check permissions for operations inside `WorkspaceService`

mhsdesign committed:
TASK: Make `pruneWorkspaceMetadata` and `pruneRoleAssignments` internal by not exposing on the `WorkspaceService`

They were introduce in #5306 for the pruning, but this is a purely internal task and there is no need for this to be api for the Neos User.
Also, the reason is that it cannot be protected with security easily.

mhsdesign commented on Neos.Neos/Classes/Domain/Service/WorkspaceService.php:
```diff
@@ -513,4 +271,20 @@ private function requireWorkspace(ContentRepositoryId $contentRepositoryId, Work
+            $this->userService->getCurrentUser()?->getId()
+        );
+        if (!$workspacePermissions->manage) {
+            throw new AccessDeniedException(sprintf('The current user does not have manage permissions for workspace "%s" in content repository "%s"', $workspaceName->value, $contentRepositoryId->value), 1731343473);
```
should we throw the flow access denied here or the CR's `\Neos\ContentRepository\Core\Feature\Security\Exception\AccessDenied` like in the other methods? Probably the latter?

mhsdesign replied:
the more i think about it, yes, otherwise some methods would throw the one and others another :D

mhsdesign commented on Neos.Neos/Classes/Domain/Service/WorkspaceService.php:
```diff
@@ -191,103 +191,31 @@ public function createPersonalWorkspaceForUserIfMissing(ContentRepositoryId $con
-        return WorkspacePermissions::create(
-            read: $userWorkspaceRole->isAtLeast(WorkspaceRole::COLLABORATOR),
-            write: $userWorkspaceRole->isAtLeast(WorkspaceRole::COLLABORATOR),
-            manage: $userIsAdministrator || $userWorkspaceRole->isAtLeast(WorkspaceRole::MANAGER),
-        );
+        return $this->metadataAndRoleRepository->getWorkspaceRoleAssignments($contentRepositoryId, $workspaceName);
```
now that we can make things `@internal` by not exposing it here on the workspace service we can consider if `getWorkspaceRoleAssignments` should really be API or if Neos can just use at this one place the  `metadataAndRoleRepository` directly? ... But probably keep it api :D

mhsdesign replied:
Why am i thinking about this? Because the

> NOTE: This should never be used to evaluate permissions, instead {@see ContentRepositoryAuthorizationService::getWorkspacePermissions()} should be used!

Makes it already seem like an internal thing somehow :D

mhsdesign replied:
But as this is to be used in the workspace ui / separate package its clearly API!

mhsdesign commented on Neos.ContentRepository.Core/Classes/Feature/Security/Exception/AccessDenied.php:
```diff
@@ -0,0 +1,34 @@
+/**
+ * @api
+ */
+final class AccessDenied extends \Exception
```
do we need some flow error option or anything so this is treated as flow `AccessDeniedException` in the parts that matter? Or does it matter at all? I do know we have some special condition sometimes to let `AccessDeniedException` always bubble up or something: (at least in the fusion runtime)?

mhsdesign replied:
@kitsunet and me discussed that this must be caught in controllers and wrapped into a flow `AccessDeniedException` to initiate a redirect or set the correct flow status to 403

kitsunet replied:
I am almost done reviewing and inclined to merge this now. we can figure out a fix for it after the fact...

kitsunet replied:
At least for typical neos ui use cases this isn't noticable as they all have controller privileges and throw the necessary exception that way already. Maybe we don't need anything for now... The frontend user case is still to be checked

mhsdesign replied:
perfect -> followup todo: https://github.com/neos/neos-development-collection/issues/5372

mhsdesign commented on Neos.Neos/Tests/Behavior/Features/Bootstrap/WorkspaceServiceTrait.php:
```diff
@@ -130,23 +131,27 @@ public function aWorkspaceWithBaseWorkspaceExistsWithoutMetadata(string $workspa
-            WorkspaceName::fromString($workspaceName),
-            WorkspaceTitle::fromString($newTitle),
-        ));
+        $this->getObject(SecurityContext::class)->withoutAuthorizationChecks(fn () =>
```
i didnt add tests yet for these things that they work _with_ security will try to do that in your followup pr :) For now its just `withoutAuthorizationChecks` :D

mhsdesign replied:
fixing it...

mhsdesign committed:
TASK: Adjust docs for privileg and permission v/o

mhsdesign committed:
TASK: inline methods in ContentRepositoryAuthorizationService instead of having them private

mhsdesign committed:
TASK: Remove obsolete warnings in docs

mhsdesign committed:
TASK: Fix cli output of `workspace:assignrole` to not throw error

mhsdesign committed:
TASK: Declare `AbstractSubtreeTagBasedPrivilege` internal as it doesnt make sense to be extended

mhsdesign reviewed (approved):
Feels already a lot better with the `WorkspaceService` being split up and the new `$roles, $userId` signature in the authorisation service. Now everything is sound :) Left a few last comments and solved the nitpicks myself. After addressing those i think this can be merged finally! Thanks for sticking to this topic so long!!!

I also tested that the Neos ui still works (e2e ran locally) and adjusted it to the latest changes.

One more thing though, maybe @kitsunet wants to have a look at the things that are evaluated often and might need to be cached? See `\Neos\Neos\Security\Authorization\ContentRepositoryAuthorizationService::getVisibilityConstraints`? And the write side now has to fetch the node again and its workspace but i guess that is _ok_ though could be cached too at some point ... idk :D

But giving my +1 already because of goodwill :D

bwaidelich committed:
Merge branch '9.0' into feature/3732-cr-privileges-v2

# Conflicts:
#	Neos.Neos/Classes/Command/CrCommandController.php

bwaidelich committed:
Merge branch '9.0' into feature/3732-cr-privileges-v2

# Conflicts:
#	Neos.ContentRepository.Core/Classes/ContentRepository.php
#	Neos.ContentRepository.Core/Classes/Factory/ContentRepositoryFactory.php
#	Neos.ContentRepositoryRegistry/Classes/ContentRepositoryRegistry.php

bwaidelich committed:
Merge branch '9.0' into feature/3732-cr-privileges-v2

bwaidelich committed:
Merge branch 'feature/3732-cr-privileges-v2' into feature/3732-cr-privileges-tests

# Conflicts:
#	Neos.Neos/Tests/Behavior/Features/Bootstrap/WorkspaceServiceTrait.php

bwaidelich committed:
Merge branch '9.0' into feature/3732-cr-privileges-v2

# Conflicts:
#	Neos.ContentRepository.Core/Classes/ContentRepository.php

bwaidelich committed:
Merge branch 'feature/3732-cr-privileges-v2' into feature/3732-cr-privileges-tests

bwaidelich committed:
Merge branch '9.0' into feature/3732-cr-privileges-v2

bwaidelich committed:
Merge branch 'feature/3732-cr-privileges-v2' into feature/3732-cr-privileges-tests

# Conflicts:
#	Neos.ContentRepository.TestSuite/Classes/Behavior/Features/Bootstrap/GenericCommandExecutionAndEventPublication.php

bwaidelich committed:
Add more tests

bwaidelich committed:
Prefix ContentRepositorySecurityTrait fields to avoid naming clashes

bwaidelich committed:
Merge pull request #5345 from neos/feature/3732-cr-privileges-tests

TASK: Tests for "Content Repository Privileges"

mhsdesign commented on Neos.Neos/Tests/Behavior/Features/Bootstrap/ContentRepositorySecurityTrait.php:
```diff
@@ -0,0 +1,210 @@
+        $policyConfiguration = ObjectAccess::getProperty($policyService, 'policyConfiguration', true);
+        $mergedPolicyConfiguration = Arrays::arrayMergeRecursiveOverrule($policyConfiguration, Yaml::parse($policies->getRaw()));
+
+        self::$crSecurity_testingPolicyPathAndFilename = $this->getObject(Environment::class)->getPathToTemporaryDirectory() . 'Policy.yaml';
+        file_put_contents(self::$crSecurity_testingPolicyPathAndFilename, Yaml::dump($mergedPolicyConfiguration));
+
+        ObjectAccess::setProperty($policyService, 'initialized', false, true);
+        $this->getObject(ConfigurationManager::class)->flushConfigurationCache();
```
is this latter part really needed? As we dont have sub processes we dont have to preserve the state and mess with the file system right?

... but ill look into it :)

bwaidelich replied:
Without that I got more failing tests because we have so many runtime caches on all layers

> ... but ill look into it :)

Great, thank you. And let me know if you need support (contentually or mentally *g)

mhsdesign committed:
TASK: Fix new Neos security tests and fake `PolicyService` correctly

- we must not write the settings to the file system via $crSecurity_testingPolicyPathAndFilename as following processes and tests in the same context would be affected. Also flushing and reloading the configuration catch (`flushConfigurationCache`) is expensive and the tests must not mess with that low level flow behaviour

- instead we only rely on faking the `PolicyService`'s runtime caches. This is done by injecting a custom configuration into `$policyConfiguration` and then initialising the `PolicyService`

mhsdesign committed:
TASK: Split `ContentRepositorySecurityTrait` and `FlowSecurityTrait`

mhsdesign committed:
TASK: Cleanup imports

mhsdesign committed:
TASK: Cleanup imports

Now we throw the cr AccessDenied exception instead of flows access denied in all places of the workspace service to make behaviour consistent.

$contentRepository->handle might for example throw the `AccessDenied` when creating a workspace

mhsdesign committed:
TASK: Remove `withoutAuthorizationChecks` in tests

And introduce the behaviour of the `FunctionalTestCase`:
https://github.com/neos/flow-development-collection/commit/b9c89e3e08649cbb5366cb769b2f79b0f13bd68e

mhsdesign committed:
TASK: Add security tests for workspace service

assert that base workspace creation is only allowed for writing

and that managing (setting title and roles) is only allowed to managers

mhsdesign commented on Neos.ContentRepository.Core/Classes/Feature/Security/Exception/AccessDenied.php:
```diff
@@ -0,0 +1,34 @@
+{
+    public static function becauseCommandIsNotGranted(CommandInterface $command, string $reason): self
+    {
+        return new self(sprintf('Command "%s" was denied: %s', $command::class, $reason), 1729086686);
```
hmm this should probably contain the cr id to ease debugging? But we have this missing alotta places in exceptions

kitsunet reviewed (approved):
Tried various scenarios locally, checked performance, looked over the code. ✅🙌👏🚀

crydotsnake commented:
<img width="806" alt="image" src="https://github.com/user-attachments/assets/b604c2e4-2475-42ab-b914-391573b8ed20" />

Mh. Normally there should be always a live workspace 😅
